python/src/array.cpp

```diff
@@ -770,37 +770,90 @@ void init_array(nb::module_& m) {
           "other"_a)
       .def(
           "__eq__",
-          [](const array& a, const ScalarOrArray v) {
+          [](const array& a,
+             const ScalarOrArray& v) -> std::variant<array, bool> {
+            auto check_for_obj = std::get_if<nb::object>(&v);
+            if (check_for_obj) {
+              // return false in case of object comparison which is not mlx
+              // array
+              if (!isMlxCoreArray(*check_for_obj)) {
+                return false;
+              }
+            }
             return equal(a, to_array(v, a.dtype()));
           },
           "other"_a)
       .def(
           "__lt__",
-          [](const array& a, const ScalarOrArray v) {
+          [](const array& a,
+             const ScalarOrArray v) -> std::variant<array, bool> {
+            auto check_for_obj = std::get_if<nb::object>(&v);
+            if (check_for_obj) {
+              // return false in case of object comparison which is not mlx
+              // array
+              if (!isMlxCoreArray(*check_for_obj)) {
+                return false;
+              }
+            }
             return less(a, to_array(v, a.dtype()));
           },
           "other"_a)
       .def(
           "__le__",
-          [](const array& a, const ScalarOrArray v) {
+          [](const array& a,
```

Comment:
Let's not enable comparison to arbitrary objects for all of these overloads, it does not make sense to do `a < "hello"`, that is fine if it raises an exception.

Reply:
Sure. But isn't it better to raise a more meaningful exception? Right now, the exception occurs in `to_array` function, while trying to do `nb::cast`. The exception is something like this:
`Cannot cast type <?> to mlx array`
I will change the exception to a more meaningful one, showing that the comparison is wrong

Reply:
I updated the code and throw this exception in these cases:
```
ValueError: Object comparison is not a valid operation on an mlx array.
```
I also added some tests to make sure it works properly.

Reply:
That's much nicer! I might go a step further and include the type name in the exception error message, that might be helfpul when debugging these.

Reply:
I see. but isn't it a little bit confusing while debugging? Because suppose one is doing this operation:
```
mlx.array([1, 2]) > [1,2]
```
and receives:
```
ValueError: Invalid type  list received in array initialization.
```
while we are not actually initialising anything here! So, the error message should somehow represent that this operation is not valid. Don't you agree?

Reply:
Indeed I see your point. But its an issue for all of our ops which take `ScalarOrArray` (which are many) so it doesn't make sense to special case these. We can try to find a way to give more useful error messaging for all cases as a follow up / separate issue.

Reply:
I think we need a more meaningful exception for all __operations__ which are invalid on an `mlx.core.array`. E.g:
```
mx.array([0.5, 0.75]) % [1, 2]
# OR
mx.array([0.5, 0.75]) * (1, 2)
# OR ...
```
Do you think this is a valid concern? and if so, how do you propose to pursue this?

Reply:
Maybe a minor concern.. I would open an issue for it so we can track / discuss, but it's beyond the scope of this PR so let's move it to an issue. We can also discuss possible fixes there.

Reply:
Sure, great!